### DISCUSSION TITLE
Special characters are wrongly encoded/decoded

### DISCUSSION THREAD

**Discussion by vpas88** [falconry/falcon#2174] [Q&A] [answered]
I have a small application using falcon-api, URI-s can have special characters like:

> íéáűúőóüö

When I send above URI, it's percent-encoded to:

> %C3%AD%C3%A9%C3%A1%C5%B1%C3%BA%C5%91%C3%B3%C3%BC%C3%B6

I use gunicorn, gunicorn pass this to falcon as:

> ÃÂ­ÃÂ©ÃÂ¡Ã

After it's encoded+decoded in request.py here (I guess):
```
if not isascii(path):
            path = path.encode('iso-8859-1').decode('utf-8', 'replace')
```

The result is:

> Â±ÃÂºÃ

Instead of the original: 

> íéáűúőóüö

**Comment by vytas7** [comment]
Hi @vpas88!
I've converted this issue to a discussion since I cannot really reproduce any of the above. By the referenced line, Falcon is merely following the [PEP 3333](https://peps.python.org/pep-3333/#the-server-gateway-side) convention to decode tunneled bytes from Latin-1.

**Comment by vytas7** [comment]
I've used the following `test.py`:
```python
import falcon


class Inspector:
    def on_get(self, req, resp, param):
        resp.media = {
            'headers': req.headers,
            'method': req.method,
            'param': param,
            'uri': req.uri,
        }


app = falcon.App()
app.add_route('/{param}', Inspector())
```

Packages used (on CPython 3.10):
```
$ pip freeze | egrep '(falcon|gunicorn)'
falcon==3.1.1
gunicorn==21.2.0
```

Sending a request similar to yours:
```
$ xh http://localhost:8000/íéáűúőóüö
HTTP/1.1 200 OK
Connection: close
Content-Length: 250
Content-Type: application/json
Server: gunicorn

{
    "headers": {
        "ACCEPT-ENCODING": "gzip, deflate, br",
        "USER-AGENT": "xh/0.18.0",
        "CONNECTION": "keep-alive",
        "ACCEPT": "*/*",
        "HOST": "localhost:8000"
    },
    "method": "GET",
    "param": "íéáűúőóüö",
    "uri": "http://localhost:8000/íéáűúőóüö"
}
```

Could it be that you had a very old version of Gunicorn, Falcon, or just your HTTP client encodes or decodes not in a way you would expect it? Which client or browser are you using?

**Comment by vpas88** [comment]
I have executed the following with your code (i used curl) and I see the same problem:
```
(venv) vpas@alderaan1:~/dev/test$ cat test.py
import falcon


class Inspector:
    def on_get(self, req, resp, param):
        resp.media = {
            'headers': req.headers,
            'method': req.method,
            'param': param,
            'uri': req.uri,
        }


app = falcon.App()
app.add_route('/{param}', Inspector())
(venv) vpas@alderaan1:~/dev/test$ gunicorn test:app test.py 
[2023-09-17 20:14:43 +0200] [1077018] [INFO] Starting gunicorn 21.2.0
[2023-09-17 20:14:43 +0200] [1077018] [INFO] Listening at: http://127.0.0.1:8000 (1077018)
[2023-09-17 20:14:43 +0200] [1077018] [INFO] Using worker: sync
[2023-09-17 20:14:43 +0200] [1077019] [INFO] Booting worker with pid: 1077019
^Z
[1]+  Stopped                 gunicorn test:app test.py
(venv) vpas@alderaan1:~/dev/test$ bg
[1]+ gunicorn test:app test.py &
(venv) vpas@alderaan1:~/dev/test$ curl http://localhost:8000/íéáűúőóüö
{"headers": {"HOST": "localhost:8000", "USER-AGENT": "curl/7.81.0", "ACCEPT": "*/*"}, "method": "GET", "param": "Ã­Ã©Ã¡Å±ÃºÅÃ³Ã¼Ã¶", "uri": "http://localhost:8000/Ã­Ã©Ã¡Å±ÃºÅÃ³Ã¼Ã¶"}
(venv) vpas@alderaan1:~/dev/test$ pip freeze | egrep '(falcon)|gunicorn'
falcon==3.1.1
falcon-apispec==0.4.0
falcon-auth==1.1.0
gunicorn==21.2.0
```

As you can see I have the same versions like you.

**Reply by CaselIT** [reply]
Hi,

Can you repoduce in a docker container, such as one of the official python ones?

Just trying to isolate the issue, by checking it it's something in your local python install

**Reply by vytas7** [reply]
It can also be not dependent on the container/Python/Falcon but related to your shell's encoding. which affects how things are decoded from terminal, and encoded to terminal.
Could you also show what your `env | grep LC_` read? And what operating system are you using?

**Reply by vpas88** [reply]
@vytas7 :

> vpas@alderaan1:\~/dev$ env | grep LC_
> LC_ADDRESS=hu_HU.UTF-8
> LC_NAME=hu_HU.UTF-8
> LC_MONETARY=hu_HU.UTF-8
> LC_PAPER=hu_HU.UTF-8
> LC_IDENTIFICATION=hu_HU.UTF-8
> LC_TELEPHONE=hu_HU.UTF-8
> LC_MEASUREMENT=hu_HU.UTF-8
> LC_TIME=hu_HU.UTF-8
> LC_NUMERIC=hu_HU.UTF-8
> vpas@alderaan1:~/dev$ 

> vpas@alderaan1:\~/dev$ lsb_release -a
> No LSB modules are available.
> Distributor ID: Ubuntu
> Description:    Ubuntu 22.04.3 LTS
> Release:        22.04
> Codename:       jammy
> vpas@alderaan1:\~/dev$ uname -a
> Linux alderaan1 6.2.0-31-generic #31~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Wed Aug 16 13:45:26 UTC 2 x86_64 x86_64 x86_64 GNU/Linux

Please note that I'm able to reproduce the same issue when I use chrome (even) from another host (so it seems to be independent from my local settings).
![image](https://github.com/falconry/falcon/assets/73040214/9d92b3c6-9a04-4082-a8dd-12baafae85b9)

(in this case gunicorn was started like: `gunicorn -b 0.0.0.0:8000 test:app test.py`)

@CaselIT : I will try it soon.

**Reply by vytas7** [reply]
It looks like it is a client issue on your side,
```python
>>> 'íéáűúőóüö'.encode('windows-1250')
b'\xed\xe9\xe1\xfb\xfa\xf5\xf3\xfc\xf6'
```
yields the same bytes as you're getting.

What happens if you enter the correct URL directly in your browser?
http://localhost:8000/%C3%AD%C3%A9%C3%A1%C5%B1%C3%BA%C5%91%C3%B3%C3%BC%C3%B6

Also try `curl -v ...` and check which URL it is using before sending it to the server.

**Reply by vpas88** [reply]
> vpas@alderaan1:~/dev/test$ cat Dockerfile 
> FROM python:latest
> COPY test.py test.py
> RUN pip install gunicorn
> RUN pip install falcon
> ENTRYPOINT ["gunicorn", "-b", "0.0.0.0:8000", "test:app", "test.py"]
> 
> vpas@alderaan1:\~/dev/test$ docker run -d -p 8000:8000 falcon-test 
> ef501ece5efe8947175816f01f82aac3bdac31c44da77974891a72ec8b22cf2e
> 
> vpas@alderaan1:\~/dev/test$ curl http://localhost:8000/íéáűúőóüö
> {"headers": {"HOST": "localhost:8000", "USER-AGENT": "curl/7.81.0", "ACCEPT": "*/*"}, "method": "GET", "param": "Ã­Ã©Ã¡Å±ÃºÅÃ³Ã¼Ã¶", "uri": "http://localhost:8000/Ã­Ã©Ã¡Å±ÃºÅÃ³Ã¼Ã¶"}
> 
> vpas@alderaan1:\~/dev/test$ curl http://localhost:8000/%C3%AD%C3%A9%C3%A1%C3%AD%C3%A9%C3%A1%C5%B1%C3%BA%C5%91%C3%B3%C3%BC%C3%B6
> {"headers": {"HOST": "localhost:8000", "USER-AGENT": "curl/7.81.0", "ACCEPT": "*/*"}, "method": "GET", "param": "íéáűúőóüö", "uri": "http://localhost:8000/íéáűúőóüö"}
>

Looks like it's working from Docker (at least when percent-encoded, which is OK)... now I'm a little bit confused what can affect the behaviour.

**Reply by vpas88** [reply]
Hm, it's okay if I use percent-encode... 

The setup is that react application is sending (axios) this data to falcon-api, so I guess the problem will be on that site, I just don't have a clue what can go wrong there? Maybe react should use ISO-8859-1 also? I don't know what is it using...

**Reply by vytas7** [reply]
I'm no expert in modern frontend, but I think you could simply give it a try. React ought to be able to handle UTF-8 correctly if your HTML page uses the correct (UTF-8) charset.

**Reply by vpas88** [reply]
OK, I think it's not react/axios. My setup is the following (in detail):
React Client -> Nginx -> Gunicorn -> Falcon

Nginx is still logging the correct percent-encoded URI, however Gunicorn is already passed "wrong" character to Falcon request.py. 
So now I'm suspecting Nginx is doing some unneeded percent-decoding...

**Reply by vytas7** [reply]
Don't think Nginx is at fault, I think it normally rejects non-ASCII URLs with a 400 :thinking:

**Reply by vpas88** [reply]
I have created a little socket server to see what NGINX send to gunicorn - now my gunicorn was that:
```
#! /usr/bin/python
# a simple tcp server
import socket,os
sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
sock.bind(('127.0.0.1', 19960))
sock.listen()
while True:
    try:
        connection,address = sock.accept()
        buf = connection.recv(1024)
        print(buf.decode())
    except KeyboardInterrupt:
        connection.close()
        break
    connection.close()
```

Result (actually it's a DELETE request, but i dont think it's important):
```
DELETE /íéáűúőóüö HTTP/1.0
Host: 127.0.0.1:19960
Connection: close
sec-ch-ua: "Google Chrome";v="117", "Not;A=Brand";v="8", "Chromium";v="117"
accept: application/json, text/plain, */*
sec-ch-ua-mobile: ?0
```

Chrome sent this (the answer was not properly handled, but you can see that it's sent percentage-encoded URI):
![image](https://github.com/falconry/falcon/assets/73040214/40624fc5-beb8-422b-974e-7f305b118762)

So my understanding is that percentage-encoded URI is properly sent from client side (React->axios->Chrome) towards NGINX where it's decoded and the utf-8(?) string sent towards gunicorn what is somehow translates this to wrong and this is sent to falcon.

**Reply by vpas88** [reply, marked as answer]
OK, finally I have manage to solve the problem. I'm putting the solution here for future reference; when using NGINX and URI is parsed in location proxy_pass will get the decoded value.
My solution was that:
```
location ~ ^/api/(.*)$ {
	rewrite  ^/api/(.*)$ $1$is_args$args break;
	proxy_pass http://127.0.0.1:19960;
}
```

This way the original percentage-encoded format is passed towards gunicorn/falcon and they will parse it properly.

**Reply by vytas7** [reply]
Good that you found it 👍

I've never had this specific issue, but had other edge cases, so in general I try not to design APIs that pass arbitrary values inside URI.

**Reply by CaselIT** [reply]
I've changed this discussion in a Q&A and marked that as the answer. thanks for the update
